indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java

```diff
@@ -500,95 +566,204 @@ public IndexTuningConfig getTuningConfig()
+    private static final boolean DEFAULT_SKIP_FIREHOSE_CACHING = false;
+
     private final FirehoseFactory firehoseFactory;
+    private final boolean appendToExisting;
+    private final boolean skipFirehoseCaching;
 
     @JsonCreator
     public IndexIOConfig(
-        @JsonProperty("firehose") FirehoseFactory firehoseFactory
+        @JsonProperty("firehose") FirehoseFactory firehoseFactory,
+        @JsonProperty("appendToExisting") @Nullable Boolean appendToExisting,
+        @JsonProperty("skipFirehoseCaching") @Nullable Boolean skipFirehoseCaching
     )
     {
       this.firehoseFactory = firehoseFactory;
+      this.appendToExisting = appendToExisting == null ? DEFAULT_APPEND_TO_EXISTING : appendToExisting;
+      this.skipFirehoseCaching = skipFirehoseCaching == null ? DEFAULT_SKIP_FIREHOSE_CACHING : skipFirehoseCaching;
     }
 
     @JsonProperty("firehose")
     public FirehoseFactory getFirehoseFactory()
     {
       return firehoseFactory;
     }
+
+    @JsonProperty("appendToExisting")
+    public boolean isAppendToExisting()
+    {
+      return appendToExisting;
+    }
+
+    @JsonProperty("skipFirehoseCaching")
+    public boolean isSkipFirehoseCaching()
+    {
+      return skipFirehoseCaching;
+    }
   }
 
   @JsonTypeName("index")
-  public static class IndexTuningConfig implements TuningConfig
+  public static class IndexTuningConfig implements TuningConfig, AppenderatorConfig
   {
     private static final int DEFAULT_TARGET_PARTITION_SIZE = 5000000;
-    private static final int DEFAULT_ROW_FLUSH_BOUNDARY = 75000;
+    private static final int DEFAULT_MAX_ROWS_IN_MEMORY = 75000;
     private static final IndexSpec DEFAULT_INDEX_SPEC = new IndexSpec();
-    private static final Boolean DEFAULT_BUILD_V9_DIRECTLY = Boolean.FALSE;
-
-    private final int targetPartitionSize;
-    private final int rowFlushBoundary;
-    private final int numShards;
+    private static final int DEFAULT_MAX_PENDING_PERSISTS = 0;
+    private static final boolean DEFAULT_BUILD_V9_DIRECTLY = false;
+    private static final boolean DEFAULT_FORCE_EXTENDABLE_SHARD_SPECS = false;
+    private static final boolean DEFAULT_REPORT_PARSE_EXCEPTIONS = false;
+
+    private final Integer targetPartitionSize;
+    private final int maxRowsInMemory;
+    private final Integer numShards;
     private final IndexSpec indexSpec;
-    private final Boolean buildV9Directly;
+    private final File basePersistDirectory;
+    private final int maxPendingPersists;
+    private final boolean buildV9Directly;
     private final boolean forceExtendableShardSpecs;
+    private final boolean reportParseExceptions;
 
     @JsonCreator
     public IndexTuningConfig(
-        @JsonProperty("targetPartitionSize") int targetPartitionSize,
-        @JsonProperty("rowFlushBoundary") int rowFlushBoundary,
+        @JsonProperty("targetPartitionSize") @Nullable Integer targetPartitionSize,
+        @JsonProperty("maxRowsInMemory") @Nullable Integer maxRowsInMemory,
```

Comment:
I also changed the semantics of how `targetPartitionSize` and `numShards` work together because I found it a bit awkward.. previously you had to set `targetPartitionSize` to `-1` if you wanted to use `numShards`; I made it so if you just leave out `targetPartitionSize` and set `numShards` it'll work, but I have a bounds check on `targetPartitionSize` to not be negative. I guess I can special case -1 for backwards compatibility.